init-tracing-opentelemetry/src/tracing_subscriber_ext.rs

```diff
@@ -77,34 +80,51 @@ pub fn regiter_otel_layers<S>(
 where
     S: Subscriber + for<'a> LookupSpan<'a>,
 {
-    let (trace_layer, tracer_provider) = build_tracer_layer()?;
-    let subscriber = subscriber.with(trace_layer);
+    register_otel_layers_with_resource(subscriber, DetectResource::default().build())
+}
 
+pub fn register_otel_layers_with_resource<S>(
+    subscriber: S,
+    otel_rsrc: Resource,
+) -> Result<(impl Subscriber + for<'span> LookupSpan<'span>, OtelGuard), Error>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
     #[cfg(feature = "metrics")]
-    {
-        let (metrics_layer, meter_provider) = build_metrics_layer()?;
-        let subscriber = subscriber.with(metrics_layer);
-        Ok((
-            subscriber,
-            OtelGuard {
-                meter_provider,
-                tracer_provider,
-            },
-        ))
-    }
-    #[cfg(not(feature = "metrics"))]
-    Ok((subscriber, OtelGuard { tracer_provider }))
+    let (metrics_layer, meter_provider) = build_metrics_layer_with_resource(otel_rsrc.clone())?;
+    let (trace_layer, tracer_provider) = build_tracer_layer_with_resource(otel_rsrc)?;
```

Comment:
The layer construction order was swapped from the previous implementation (trace -> metrics) due to my perfectionism about minimizing clones.